segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java

```diff
@@ -0,0 +1,1310 @@
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+/**
+ * Implements storage for segments using {@link ChunkStorage} and {@link ChunkMetadataStore}.
+ * The metadata about the segments is stored in metadataStore using two types of records {@link SegmentMetadata} and {@link ChunkMetadata}.
+ * Any changes to layout must be made inside a {@link MetadataTransaction} which will atomically change the records upon
+ * {@link MetadataTransaction#commit()}.
+ * Detailed design is documented here https://github.com/pravega/pravega/wiki/PDP-34:-Simplified-Tier-2
+ */
+@Slf4j
+@Beta
+public class ChunkedSegmentStorage implements Storage {
+    /**
+     * Configuration options for this ChunkedSegmentStorage instance.
+     */
+    @Getter
+    private final ChunkedSegmentStorageConfig config;
+
+    /**
+     * Metadata store containing all storage data.
+     * Initialized by segment container via {@link ChunkedSegmentStorage#bootstrap(int, ChunkMetadataStore)}.
+     */
+    @Getter
+    private ChunkMetadataStore metadataStore;
+
+    /**
+     * Underlying {@link ChunkStorage} to use to read and write data.
+     */
+    @Getter
+    private final ChunkStorage chunkStorage;
+
+    /**
+     * Storage executor object.
+     */
+    private final Executor executor;
+
+    /**
+     * Tracks whether this instance is closed or not.
+     */
+    private final AtomicBoolean closed;
+
+    /**
+     * Current epoch of the {@link Storage} instance.
+     * Initialized by segment container via {@link ChunkedSegmentStorage#initialize(long)}.
+     */
+    @Getter
+    private long epoch;
+
+    /**
+     * Id of the current Container.
+     * Initialized by segment container via {@link ChunkedSegmentStorage#bootstrap(int, ChunkMetadataStore)}.
+     */
+    @Getter
+    private int containerId;
+
+    /**
+     * {@link SystemJournal} that logs all changes to system segment layout so that they can be are used during system bootstrap.
+     */
+    @Getter
+    private SystemJournal systemJournal;
+
+    /**
+     * {@link ReadIndexCache} that has index of chunks by start offset
+     */
+    private final ReadIndexCache readIndexCache;
+
+    /**
+     * List of garbage chunks.
+     */
+    private final List<String> garbageChunks = new ArrayList<String>();
+
+    /**
+     * Prefix string to use for logging.
+     */
+    private String logPrefix;
+
+    /**
+     * Creates a new instance of the ChunkedSegmentStorage class.
```

Comment:
Some classes in the Javadoc are mentioned in plain text, whereas others are using `{@link xxx}` . I think that the latter is the right one.

Reply:
fixed.